cartography/intel/gsuite/api.py

```diff
@@ -0,0 +1,243 @@
+            users = []
+            break
+        users = users + resp.get('users', [])
+        # break
+        request = admin.users().list_next(request, resp)
+    return users
+
+
+def get_ingestion_groups_qry():
+    return """
+        UNWIND {GroupData} as group
+        MERGE (g:GSuiteGroup{id: group.id})
+        ON CREATE SET
+        g.admin_created = group.adminCreated,
+        g.description = group.description,
+        g.direct_members_count = group.directMembersCount,
+        g.email = group.email,
+        g.etag = group.etag,
+        g.kind = group.kind,
+        g.name = group.name,
+        g.lastupdated = {UpdateTag}
+    """
+
+
+def load_gsuite_groups(session, groups, gsuite_update_tag):
+    logger.info('Ingesting {} gsuite groups'.format(len(groups)))
+    session.run(get_ingestion_groups_qry(), GroupData=groups, UpdateTag=gsuite_update_tag)
+
+
+def get_ingestion_users_qry():
+    return """
+        UNWIND {UserData} as user
+        MERGE (u:GSuiteUser{id: user.id})
+        ON CREATE SET
+        u.agreed_to_terms = user.agreedToTerms,
+        u.archived = user.archived,
+        u.change_password_at_next_login = user.changePasswordAtNextLogin,
+        u.creation_time = user.creationTime,
+        u.customer_id = user.customerId,
+        u.etag = user.etag,
+        u.include_in_global_address_list = user.includeInGlobalAddressList,
+        u.ip_whitelisted = user.ipWhitelisted,
+        u.is_admin = user.isAdmin,
+        u.is_delegated_admin =  user.isDelegatedAdmin,
+        u.is_enforced_in_2_sv = user.isEnforcedIn2Sv,
+        u.is_enrolled_in_2_sv = user.isEnrolledIn2Sv,
+        u.is_mailbox_setup = user.isMailboxSetup,
+        u.kind = user.kind,
+        u.last_login_time = user.lastLoginTime,
+        u.name = user.name.fullName,
+        u.family_name = user.name.familyName,
+        u.given_name = user.name.givenName,
+        u.org_unit_path = user.orgUnitPath,
+        u.primary_email = user.primaryEmail,
+        u.suspended = user.suspended,
+        u.thumbnail_photo_etag = user.thumbnailPhotoEtag,
+        u.thumbnail_photo_url = user.thumbnailPhotoUrl,
+        u.lastupdated = {UpdateTag}
+    """
+
+
+def load_gsuite_users(session, users, gsuite_update_tag):
+    logger.info('Ingesting {} gsuite users'.format(len(users)))
+    session.run(get_ingestion_users_qry(), UserData=users, UpdateTag=gsuite_update_tag)
+
+
+def get_ingestion_membership_qry():
+    """
+    MATCH (user:GSuiteUser),(group:GSuiteGroup)
+        WHERE user.id = '{UserID}' AND group.id = '{GroupID}'
+        MERGE (user)-[r:MEMBER_GSUITE_GROUP]->(group)
+        ON CREATE SET r.lastupdated = {UpdateTag}
+    """
+    return """
+        UNWIND {MemberData} as member
+        MATCH (user:GSuiteUser {id: member.id}),(group:GSuiteGroup {id: {GroupID} })
+        MERGE (user)-[r:MEMBER_GSUITE_GROUP]->(group)
+    """
+
+
+def load_gsuite_members(session, group, members, gsuite_update_tag):
```

Comment:
Need an extra pair of eyes on this function.  It's doing two queries to link:
1.  users to groups
2.  groups to groups (since a group can have a group)

Not sure if there's a more performant way of doing this.

Reply:
Two separate queries is perfectly fine.